packages/@glimmer/runtime/lib/vm/rehydrate-builder.ts

```diff
@@ -72,55 +71,85 @@ export class RehydrateBuilder extends NewElementBuilder implements ElementBuilde
-         *     <!--%+b:2%--> <-  currentCursor.candidate.firstChild -> cursor.candidate
-         *     Foo
-         *     <!--%-b:2%-->
-         *   </div>
-         *   <!--%-b:1%-->  <--  becomes currentCursor.candidate
-         */
-
-        // where to rehydrate from if we are in rehydration mode
-        cursor.candidate = element.firstChild;
-        // where to continue when we pop
-        currentCursor.candidate = element.nextSibling;
-      }
+  disableRehydration(nextSibling: Option<SimpleNode>) {
+    let currentCursor = this.currentCursor!;
+
+    // rehydration will be disabled until we either:
+    // * hit popElement (and return to using the parent elements cursor)
+    // * hit closeBlock and the next sibling is a close block comment
+    //   matching the expected openBlockDepth
+    currentCursor.candidate = null;
+    currentCursor.nextSibling = nextSibling;
+  }
+
+  enableRehydration(candidate: Option<SimpleNode>) {
+    let currentCursor = this.currentCursor!;
+
+    currentCursor.candidate = candidate;
+    currentCursor.nextSibling = null;
+  }
+
+  pushElement(
+    /** called from parent constructor before we initialize this */
+    this:
+      | RehydrateBuilder
+      | (NewElementBuilder & Partial<Pick<RehydrateBuilder, 'blockDepth' | 'candidate'>>),
+    element: SimpleElement,
+    nextSibling: Maybe<SimpleNode> = null
+  ) {
+    let cursor = new RehydratingCursor(element, nextSibling, this.blockDepth || 0);
+
+    /**
+     * <div>   <---------------  currentCursor.element
+     *   <!--%+b:1%--> <-------  would have been removed during openBlock
+     *   <div> <---------------  currentCursor.candidate -> cursor.element
+     *     <!--%+b:2%--> <-----  currentCursor.candidate.firstChild -> cursor.candidate
+     *     Foo
+     *     <!--%-b:2%-->
+     *   </div>
+     *   <!--%-b:1%-->  <------  becomes currentCursor.candidate
+     */
+    if (this.candidate !== null) {
+      cursor.candidate = element.firstChild;
+      this.candidate = element.nextSibling;
     }
+
     this[CURSOR_STACK].push(cursor);
   }
 
+  // clears until the end of the current container
+  // either the current open block or higher
   private clearMismatch(candidate: SimpleNode) {
     let current: Option<SimpleNode> = candidate;
     let currentCursor = this.currentCursor;
     if (currentCursor !== null) {
       let openBlockDepth = currentCursor.openBlockDepth;
       if (openBlockDepth >= currentCursor.startingBlockDepth) {
-        while (current && !(isComment(current) && getCloseBlockDepth(current) === openBlockDepth)) {
+        while (current) {
+          if (isCloseBlock(current)) {
+            let closeBlockDepth = getBlockDepth(current);
+            if (openBlockDepth >= closeBlockDepth) {
+              // cleared up until the close but we haven't closed the current
+              // block unless we are above
+              currentCursor.openBlockDepth = closeBlockDepth;
```

Comment:
remove this line, and I will describe the unit test tomorrow

Reply:
Removed, will try to review with you in the morning to figure out the failing test case...